sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -269,4 +272,60 @@ case class InsertIntoHadoopFsRelationCommand(
       }
     }.toMap
   }
+
+  private def isSubDir(
+      src: Path,
+      dest: Path,
+      hadoopConf: Configuration): Boolean = {
+    if (src == null) {
+      false
+    } else {
+      val srcFs = src.getFileSystem(hadoopConf)
+      val destFs = dest.getFileSystem(hadoopConf)
+      val fullSrc = srcFs.makeQualified(src).toString + Path.SEPARATOR
+      val fullDest = destFs.makeQualified(dest).toString + Path.SEPARATOR
+      val schemaSrcf = src.toUri.getScheme
+      val schemaDestf = dest.toUri.getScheme
+      if (schemaSrcf != null && schemaDestf != null && !(schemaSrcf == schemaDestf)) {
+        false
+      } else {
+        fullSrc.startsWith(fullDest)
+      }
+    }
+  }
+
+  def checkWritePathReadFrom(
```

Comment:
how expensive is it?

Reply:
> how expensive is it?

If use `FileSourceScan.dynamicallySelectedPartitions`, it  might have to wait for the expression to be evaluated before making a judgment, which might cause the SQL that could not execute properly to execute partially before throwing an exception.  But this is the most accurate target partition to get.

If use `FileSourceScan.selectedPartitions`, the cost is cheap, since there is no need for dynamic partitioning to push down the result, there is no need to wait for the actual execution.

Reply:
> how expensive is it?

IMO, use `FileSourceScan.selectedPartitions` maybe enough。